Thanks for the reviews @simeoncarstens & @SuperSandro2000.

Aesara's tests fail with `impure path [...] used in link`:
```
============================= test session starts ==============================
platform linux -- Python 3.9.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /build/source, configfile: setup.cfg, testpaths: tests/
collected 5269 items / 1 error / 28 skipped / 5240 selected                    

==================================== ERRORS ====================================
______________ ERROR collecting tests/tensor/random/test_basic.py ______________
tests/tensor/random/test_basic.py:662: in <module>
    (at.ones((M_at, M_at + 1)), ()),
aesara/tensor/basic.py:1002: in ones
    return alloc(np.array(1, dtype=dtype), *shape)
aesara/graph/op.py:283: in __call__
    node = self.make_node(*inputs, **kwargs)
aesara/tensor/basic.py:1502: in make_node
    sh, bcast = infer_broadcastable(shape)
aesara/tensor/basic.py:1469: in infer_broadcastable
    folded_shape = optimize_graph(shape_fg, custom_opt=topo_constant_folding).outputs
aesara/graph/opt_utils.py:47: in optimize_graph
    _ = canonicalize_opt.optimize(fgraph)
aesara/graph/opt.py:101: in optimize
    ret = self.apply(fgraph, *args, **kwargs)
aesara/graph/opt.py:289: in apply
    self.failure_callback(e, self, optimizer)
aesara/graph/opt.py:224: in warn
    raise exc
aesara/graph/opt.py:278: in apply
    sub_prof = optimizer.optimize(fgraph)
aesara/graph/opt.py:101: in optimize
    ret = self.apply(fgraph, *args, **kwargs)
aesara/graph/opt.py:2504: in apply
    sub_prof = gopt.apply(fgraph)
aesara/graph/opt.py:2107: in apply
    nb += self.process_node(fgraph, node)
aesara/graph/opt.py:2000: in process_node
    self.failure_callback(
aesara/graph/opt.py:1903: in warn_inplace
    return NavigatorOptimizer.warn(exc, nav, repl_pairs, local_opt, node)
aesara/graph/opt.py:1891: in warn
    raise exc
aesara/graph/opt.py:1997: in process_node
    replacements = lopt.transform(fgraph, node)
aesara/graph/opt.py:1204: in transform
    return self.fn(fgraph, node)
aesara/tensor/basic_opt.py:2951: in constant_folding
    thunk = node.op.make_thunk(node, storage_map, compute_map, no_recycling=[])
aesara/graph/op.py:690: in make_thunk
    return self.make_c_thunk(node, storage_map, compute_map, no_recycling)
aesara/graph/op.py:656: in make_c_thunk
    outputs = cl.make_thunk(
aesara/link/c/basic.py:1204: in make_thunk
    cthunk, module, in_storage, out_storage, error_storage = self.__compile__(
aesara/link/c/basic.py:1139: in __compile__
    thunk, module = self.cthunk_factory(
aesara/link/c/basic.py:1635: in cthunk_factory
    module = get_module_cache().module_from_key(key=key, lnk=self)
aesara/link/c/cmodule.py:1198: in module_from_key
    module = lnk.compile_cmodule(location)
aesara/link/c/basic.py:1544: in compile_cmodule
    module = c_compiler.compile_str(
aesara/link/c/cmodule.py:2553: in compile_str
    raise CompileError(
E   aesara.link.c.exceptions.CompileError: Compilation failed (return status=1):
E   /nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/g++ -shared -g -O3 -fno-math-errno -Wno-unused-label -Wno-unused-variable -Wno-write-strings -mtune=generic -march=x86-64 -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -m64 -fPIC -I/nix/store/hzi6fdch87l6hx9yyb6c3c2qbapczpg0-python3.9-numpy-1.21.5/lib/python3.9/site-packages/numpy/core/include -I/nix/store/afi0ysqw20yiiw2gr2d28dx40bc4ddf8-python3-3.9.10/include/python3.9 -I/build/source/aesara/link/c/c_code -L/nix/store/afi0ysqw20yiiw2gr2d28dx40bc4ddf8-python3-3.9.10/lib -fvisibility=hidden -o /homeless-shelter/.aesara/compiledir_Linux-5.4--generic-x86_64-with-glibc2.33--3.9.10-64/tmpl3kgilx3/mfd7e6d3547b91eb16dd42369ba7533d8a8769a2ee66a413b9ed97a60305d8231.so /homeless-shelter/.aesara/compiledir_Linux-5.4--generic-x86_64-with-glibc2.33--3.9.10-64/tmpl3kgilx3/mod.cpp -lpython3.9
E   impure path `/homeless-shelter/.aesara/compiledir_Linux-5.4--generic-x86_64-with-glibc2.33--3.9.10-64/tmpl3kgilx3/mfd7e6d3547b91eb16dd42369ba7533d8a8769a2ee66a413b9ed97a60305d8231.so' used in link
E   collect2: error: ld returned 1 exit status
```
</p>
</details>


About which the [NixOS wiki](https://nixos.wiki/wiki/Development_environment_with_nix-shell) says the following:

> **Troubleshooting**
> 
> When compiling software which links against local files (e.g. when compiling with rust's cargo), you may encounter the following problem:
> ```
> = note: impure path `/[...]' used in link
> ```
> This happens due to a specialty in nix: ld is wrapped in a shell script which refuses to link against files not residing in the nix store, to ensure the purity of builds. Obviously this is not useful when building locally, for example in your home directory. To disable this behavior simply set
> ```
> NIX_ENFORCE_PURITY=0
> ```
> in the nix-shell. 

So the tests run fine locally when I don't enforce purity with `NIX_ENFORCE_PURITY=0`, but I'm not sure how to handle it for this PR.

--- a/pkgs/development/python-modules/cons/default.nix
+++ b/pkgs/development/python-modules/cons/default.nix
@@ -1,0 +1,43 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, logical-unification
+, pytestCheckHook
+, pytest-html
+}:
+
+buildPythonPackage rec {
+  pname = "cons";
+  version = "0.4.5";
+
+  src = fetchFromGitHub {
+    owner = "pythological";
+    repo = "python-cons";
+    rev = "fbeedfc8a3d1bff4ba179d492155cdd55538365e";
+    sha256 = "sha256-ivHFep9iYPvyiBIZKMAzqrLGnQkeuxd0meYMZwZFFH0=";
+  };
+
+  propagatedBuildInputs = [
+    logical-unification
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    pytest-html
+  ];
+
+  pytestFlagsArray = [
+    "--html=testing-report.html"
+    "--self-contained-html"
+  ];
+
+  pythonImportsCheck = [ "cons" ];
+
+  meta = with lib; {
+    description = "An implementation of Lisp/Scheme-like cons in Python";
+    homepage = "https://github.com/pythological/python-cons";
+    changelog = "https://github.com/pythological/python-cons/releases";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ Etjean ];
+  };
+}

--- a/pkgs/development/python-modules/etuples/default.nix
+++ b/pkgs/development/python-modules/etuples/default.nix
@@ -1,0 +1,45 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, cons
+, multipledispatch
+, pytestCheckHook
+, pytest-html
+}:
+
+buildPythonPackage rec {
+  pname = "etuples";
+  version = "0.3.4";
+
+  src = fetchFromGitHub {
+    owner = "pythological";
+    repo = "etuples";
+    rev = "35d760ceb64ec318f302a6e4d3a4a80feda97a9e";
+    sha256 = "sha256-CXD8MhsdWYAcG5WDVTT/A2HDtiO1xfQbrwlYVnxXpBU=";
+  };
+
+  propagatedBuildInputs = [
+    cons
+    multipledispatch
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    pytest-html
+  ];
+
+  pytestFlagsArray = [
+    "--html=testing-report.html"
+    "--self-contained-html"
+  ];
+
+  pythonImportsCheck = [ "etuples" ];
+
+  meta = with lib; {
+    description = "Python S-expression emulation using tuple-like objects";
+    homepage = "https://github.com/pythological/etuples";
+    changelog = "https://github.com/pythological/etuples/releases";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ Etjean ];
+  };
+}

--- a/pkgs/development/python-modules/logical-unification/default.nix
+++ b/pkgs/development/python-modules/logical-unification/default.nix
@@ -1,0 +1,48 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, toolz
+, multipledispatch
+, pytestCheckHook
+, pytest-html
+, pytest-benchmark
+}:
+
+buildPythonPackage rec {
+  pname = "logical-unification";
+  version = "0.4.5";
+
+  src = fetchFromGitHub {
+    owner = "pythological";
+    repo = "unification";
+    rev = "707cf4a39e27a4a8bf06b7e7dce7223085574e65";
+    sha256 = "sha256-3wqO0pWWFRQeoGNvbSDdLNYFyjNnv+O++F7+vTBUJoI=";
+  };
+
+  propagatedBuildInputs = [
+    toolz
+    multipledispatch
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    pytest-html
+    pytest-benchmark  # Needed for the `--benchmark-skip` flag
+  ];
+
+  pytestFlagsArray = [
+    "--benchmark-skip"
+    "--html=testing-report.html"
+    "--self-contained-html"
+  ];
+
+  pythonImportsCheck = [ "unification" ];
+
+  meta = with lib; {
+    description = "Straightforward unification in Python that's extensible via generic functions";
+    homepage = "https://github.com/pythological/unification";
+    changelog = "https://github.com/pythological/unification/releases";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ Etjean ];
+  };
+}

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -2655,6 +2655,8 @@ in {
 
   eth-utils = callPackage ../development/python-modules/eth-utils { };
 
+  etuples = callPackage ../development/python-modules/etuples { };
+
   et_xmlfile = callPackage ../development/python-modules/et_xmlfile { };
 
   ev3dev2 = callPackage ../development/python-modules/ev3dev2 { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -1807,6 +1807,8 @@ in {
 
   connexion = callPackage ../development/python-modules/connexion { };
 
+  cons = callPackage ../development/python-modules/cons { };
+
   consonance = callPackage ../development/python-modules/consonance { };
 
   constantly = callPackage ../development/python-modules/constantly { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -4759,6 +4759,8 @@ in {
 
   logfury = callPackage ../development/python-modules/logfury { };
 
+  logical-unification = callPackage ../development/python-modules/logical-unification { };
+
   logilab_astng = callPackage ../development/python-modules/logilab_astng { };
 
   logilab_common = callPackage ../development/python-modules/logilab/common.nix { };

